### ISSUE TITLE
3D CNNs based approach

### ISSUE THREAD

**Issue by gsaurabhr** [csndl-iitd/realtime-sleep-staging#10]
- [x] Dataset
- [ ] Basic visualization (follow #7 )
- [ ] Literature
    - [x] #8 
    - [x] @Tanvig Add the other paper for which you found code here
- [ ] Implementing the (2+1)D model (used in both papers above)
- [ ] Implementing the 3D model (used in both papers above)

**Comment by gsaurabhr** [comment]
Dataset needs to have high density EEG. We found only one such dataset: https://github.com/csndl-iitd/realtime-sleep-staging/issues/5#issuecomment-2126712946

**Comment by gsaurabhr** [comment]
Basic data preprocessing:

1. Down sample to 100 Hz
2. Apply band pass filter between 0.2 Hz and 40 Hz (which also removes all line noise)
3. Re-reference to mastoid electrodes (what are the electrode names?)
4. Epoch the data
    - Use the annotations to detect 30 second epochs that are manually labeled
    - Generate individual .npy files for each epoch with all 64 electrode pre-processed data and save with appropriate filename

**Comment by Tanvig** [comment]
### Paper
[Tran, D., Wang, H., Torresani, L., Ray, J., LeCun, Y., & Paluri, M. (2018). A closer look at spatiotemporal convolutions for action recognition. In Proceedings of the IEEE conference on Computer Vision and Pattern Recognition (pp. 6450-6459).](https://arxiv.org/pdf/1711.11248)

### GitHub repositories:

- https://github.com/leftthomas/R2Plus1D-C3D 
- https://github.com/murari023/R2Plus1D 
- https://github.com/facebookresearch/VMZ 
- https://github.com/murilovarges/R2Plus1D 

### Summary:

- The study is motivated by the observation that 2D CNNs applied to individual video frames perform well in action recognition. It demonstrated that 3D CNNs surpass 2D CNNs within the residual learning framework. Furthermore, it showed that decomposing 3D convolutional filters into separate spatial and temporal components using the **R(2+1)D model**, based on ResNet architecture, significantly improves accuracy.

- They tested various models, such as 2D convolutions (R2D and f-R2D), 3D convolutions (R3D), Mixed convolutions (MCx and rMCx) and R(2+1)D convolutions. 

- The **2D convolutions** treat each frame independently and fail to capture temporal information adequately.

- The **R3D convolution** captures spatial and temporal information jointly but is computationally expensive and less effective than decomposed models.

- The **mixed models** combine 2D and 3D convolutions within the same network, mixing them at various layers to balance spatial and temporal feature extraction. These may not fully exploit the temporal information available in the video sequences, leading to suboptimal performance, especially on tasks requiring long-term context understanding.

- The **R(2+1)D model** decomposes a 3D convolution into a sequence of 2D and 1D convolutions. A 2D convolution of 1×7×7 (handling spatial dimensions) is followed by a 1D convolution of 3×1×1 (handling the temporal dimension) across consecutive frames. It **consistently outperformed** all the above models across various datasets.

- Compared to full 3D convolution, (2+1)D decomposition offers two advantages:

     - **Increased Nonlinearities**: It doubles the number of nonlinearities in the network due to the additional ReLU between the 2D and 1D convolutions in each block, increasing the complexity of representable functions.
   
     - **Easier Optimization**: Forcing 3D convolution into separate spatial and temporal components simplifies optimization, resulting in lower training error compared to 3D convolutional networks of the same capacity. The gap in training losses is larger for deeper networks, indicating the facilitation in optimization increases with depth.

**Comment by Tanvig** [comment]
The below GitHub repositories use **3D-CNN for EEG data**. They might be useful when we convert EEG data into image frames for our model.

- https://github.com/KvanNoord/3D-CNN-EEG-Emotion-Classification

- https://github.com/Daisybiubiubiu/EEG-Emotion-Recognition

- https://github.com/AkiKono/EEG-3D-CNN

- https://github.com/matt-houk/MB3DCNN

**Comment by Tanvig** [comment]
# #7 Basic Data Visualizations


## Data information:

<meta charset="utf-8"><b style="font-weight:normal;" id="docs-internal-guid-eac8e0b5-7fff-c789-152e-23b02cde6d4a"><div dir="ltr" style="margin-left:0pt;" align="left">
Dataset Description | Information | Comments
-- | -- | --
Number of subjects | 19 | Performed two different cognitive tasks on two different days before napping. Link: https://osf.io/zcu2w
Number of recordings | 36 | (2-night recordings of 17 subjects and 1-night recording of 2 subjects) Link: https://github.com/nmningmei/Get_Sleep_data/blob/main/data/available_subjects.csv
Number of channels | 64 | 62 EEG + 2 EOG Link: https://osf.io/ebvsr
Original sampling frequency | 1000 Hz |  
Original highpass and lowpass filters | highpass: 0.0 Hz lowpass: 500.0 Hz |  

</div></b>

## Data preprocessing steps:

- Downsampling to 100 Hz.

- Applying bandpass filter between 0.2 Hz and 40 Hz (this also removes all line noise around 60 Hz).

- Performing average re-referencing on the EEG raw data. Alternatively, we can also perform mastoid re-referencing using TP9 and TP10 electrode positions (refer [here](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5390022/)).

## Visualizations
The below visualizations are for single subject 29 (day 1)

This notebook contains all the visualizations and details: https://colab.research.google.com/drive/1QYWn7DLtCCCWRf5erdWIgdv6hdFR9-xH?usp=sharing 

### Before pre-processing

- The raw EEG plots are too heavy to be uploaded here

- For simplicity, this is a plot for 'Oz' channel for 30-sec duration (X -axis: Time (ms), Y-axis: Voltage (uV))

![image](https://github.com/csndl-iitd/realtime-sleep-staging/assets/7419416/6cce2867-2f54-4d22-a17d-229e87545cc9)

- Power Spectral Density (PSD) plot for all the channels

![image](https://github.com/csndl-iitd/realtime-sleep-staging/assets/7419416/5e1875a1-165a-4593-9284-0612ef333807)

### Hypnogram

![image](https://github.com/csndl-iitd/realtime-sleep-staging/assets/7419416/db2e715e-77a4-4be7-ab8a-c6c008a32169)

### After pre-processing

- For comparison , this is a plot for 'Oz' channel for 30-sec duration (X -axis: Time (ms), Y-axis: Voltage (uV))

![image](https://github.com/csndl-iitd/realtime-sleep-staging/assets/7419416/73570d55-29e8-42ea-8b27-5faff6a43a71)

- Power Spectral Density (PSD) plot for all the channels. We can see that the line noise at 60 Hz has disappeared due to bandpass filtering. We can also see a notch around 12-13 Hz which may signify the alpha waves with high power.

![image](https://github.com/csndl-iitd/realtime-sleep-staging/assets/7419416/06704688-5168-481d-a784-f831765caa40)


- Below is a spectrogram for a single channel (Oz). The x-axis ranges from 0 - 1804.95 secs. (i.e. the total duration).

![image](https://github.com/csndl-iitd/realtime-sleep-staging/assets/7419416/5b8707ec-f417-47ee-b01f-438afff0ee0d)

The below plot shows the spectrogram for single channel Oz for a duration of 30 secs from starting. We can observe the high power spectrum around 11-12 Hz (might be alpha waves)

![image](https://github.com/csndl-iitd/realtime-sleep-staging/assets/7419416/93c46bfe-d73d-4cec-accb-14bec508bef4)

**Comment by Tanvig** [comment]
### tSNE plot for all subjects

The below tSNE plots were created using 30 seconds epochs from all the subjects. These epochs are given as input features to the tSNE plots. The data is z-scaled before plotting. The three plots show varying level of perplexity (parameter relating to the number of nearest neighbors).

![tsne_all_subjects_perplexity30](https://github.com/csndl-iitd/realtime-sleep-staging/assets/7419416/90ccb8c3-b234-418a-bd25-6447b3ab5daa)

![tsne_all_subjects_perplexity50](https://github.com/csndl-iitd/realtime-sleep-staging/assets/7419416/2b7c97fb-a9dc-45c0-a11b-fc20390989ed)

![tsne_all_subjects_perplexity75](https://github.com/csndl-iitd/realtime-sleep-staging/assets/7419416/678afb62-4f58-43e6-bae7-8f073b5b4d80)

Below are the vaious snapshots of the 3D tSNE plot for all the subjects

![3d-tsne_1](https://github.com/csndl-iitd/realtime-sleep-staging/assets/7419416/dca0a8c1-75ad-45ed-8a9c-8e6d43cdd2d0)

![3d_tsne_2](https://github.com/csndl-iitd/realtime-sleep-staging/assets/7419416/bf630b58-6058-4234-b612-73f949a479de)

![3d_tsne_3](https://github.com/csndl-iitd/realtime-sleep-staging/assets/7419416/6bb7fe9f-9b01-4ca4-81da-18f5e8aebb3d)


### UMAP plots for all subjects

The below UMAP plots were created using 30 seconds epochs from all the subjects. These epochs are given as input features to the UMAP plots. The data is z-scaled before plotting. The two plots show varying levels of n_neighbours (parameter that balances local versus global structure in the data). The last one shows 3 dimensional plot.

![UMAP_all_subjects_neighbours15](https://github.com/csndl-iitd/realtime-sleep-staging/assets/7419416/e39e0e9a-a7ad-4684-be77-3d6dd76d2f64)

![UMAP_all_subjects_neighbours8](https://github.com/csndl-iitd/realtime-sleep-staging/assets/7419416/62d52394-0b41-48b0-b536-13b396aa8dbc)

![3d_umap](https://github.com/csndl-iitd/realtime-sleep-staging/assets/7419416/458863c8-34dc-4dd4-ade5-7882cb04cce1)

**Comment by Tanvig** [comment]
Project update:
